File: Silver/2016/US Open/03_closing_the_farm/ClosingTheFarm.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.PrintWriter;
import java.util.*;

/**
 * @author Red
 * @version 1.0
 */
public class ClosingTheFarm {
    static int n;
    static int m;
    static List<List<Integer>> graph = new ArrayList<>();
    static boolean[] closed;
    static boolean[] visited;
    static int[] closeOrder;
    static int visitedCount;

    public static void main(String[] args) throws Exception {
        BufferedReader r = new BufferedReader(new FileReader("closing.in"));
        PrintWriter pw = new PrintWriter("closing.out");
        StringTokenizer st = new StringTokenizer(r.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        closeOrder = new int[n];
        closed = new boolean[n];
        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(r.readLine());
            int a = Integer.parseInt(st.nextToken()) - 1;
            int b = Integer.parseInt(st.nextToken()) - 1;
            graph.get(a).add(b);
            graph.get(b).add(a);
        }
        for (int i = 0; i < n; i++) {
            closeOrder[i] = Integer.parseInt(r.readLine()) - 1;
        }
        // 依次关闭barn，然后从最后一个barn开始dfs查看图的连通性
        for (int i = 0; i < n; i++) {
            visited = new boolean[n];
            visitedCount = 0;
            dfs(closeOrder[n - 1]);
            if (visitedCount == n - i) {
                pw.println("YES");
            } else {
                pw.println("NO");
            }
            closed[closeOrder[i]] = true;
        }
        pw.close();
    }

    /**
     * 深度优先搜索遍历图
     */
    static void dfs(int node) {
        if (visited[node] || closed[node]) {
            return;
        }
        visitedCount++;
        visited[node] = true;
        for (int neighbor : graph.get(node)) {
            dfs(neighbor);
        }
    }

}

File: Silver/2016/US Open/03_closing_the_farm/closing_the_farm.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>

using namespace std;

int n;
int m;
vector<vector<int>> graph;
vector<bool> closed;
vector<bool> visited;
vector<int> closeOrder;
int visitedCount;
// 深度优先搜索遍历图
void dfs(int node) {
    if (visited[node] || closed[node]) {
        return;
    }

    visitedCount++;
    visited[node] = true;

    for (int neighbor : graph[node]) {
        dfs(neighbor);
    }
}

int main() {
    ifstream fin("closing.in");
    ofstream fout("closing.out");

    fin >> n >> m;

    closeOrder.resize(n);
    closed.resize(n, false);
    graph.resize(n, vector<int>());

    for (int i = 0; i < n; i++) {
        graph[i].resize(0);
    }

    for (int i = 0; i < m; i++) {
        int a, b;
        fin >> a >> b;
        a--; b--;
        graph[a].push_back(b);
        graph[b].push_back(a);
    }

    for (int i = 0; i < n; i++) {
        fin >> closeOrder[i];
        closeOrder[i]--;
    }
    // 依次关闭barn，然后从最后一个barn开始dfs查看图的连通性
    for (int i = 0; i < n; i++) {
        visited = vector<bool>(n, false);
        visitedCount = 0;
        dfs(closeOrder[n - 1]);

        if (visitedCount == n - i) {
            fout << "YES" << endl;
        } else {
            fout << "NO" << endl;
        }

        closed[closeOrder[i]] = true;
    }

    fout.close();

    return 0;
}

